PP-1335: PTL test case fails before setUp function with AttributeError

shilpakodli opened the pull request:
#### Bug/feature Description
* [PP-1335](https://pbspro.atlassian.net/browse/PP-1335)
* PTL throws AttributeError if a test suite fails before calling "setUp".
Error:
 File "/home/pbsroot/TEST/tmp/ptl/lib/python2.7/site-packages/ptl/utils/plugins/ptl_report_json.py", line 69, in get_json
    'run_id': data['start_time'].strftime('%s'),
AttributeError: 'int' object has no attribute 'strftime'

#### Affected Platform(s)
* This issue is seen only on Cray platform as it was booted to a different image other than default one which didn't have apstat installed on it.
Now this issue is not reproducible but still PTL has bug. 

#### Cause / Analysis / Design
*  When a test case fails before calling setUp method, PTL throws error while creating json report. This is due to start_time not being initialized.
* Currently start_time is assigned to 0 by default and it gets assigned when test case starts executing.
* If start_time gets assigned then __get_timeout throws error with conf error:
File "/tmp/PTL/lib/python2.7/site-packages/ptl/utils/plugins/ptl_test_runner.py", line 599, in startTest
timeout = self.__get_timeout(test)
File "/tmp/PTL/lib/python2.7/site-packages/ptl/utils/plugins/ptl_test_runner.py", line 551, in __get_timeout
__conf = getattr(test.test, 'conf')
AttributeError: 'Failure' object has no attribute 'conf'

#### Solution Description
*  Initialized start_time to datetime.datetime.now() by default before test case starts executing*
*  modify __get_timeout to return default timeout value if there is any error in except block.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

hirenvadalia commented on test/fw/ptl/utils/plugins/ptl_test_db.py:
```diff
@@ -1625,6 +1626,9 @@ def write(self, data, logfile=None):
         if len(data) == 0:
             return
         if 'testdata' in data.keys():
+            if (data['testdata']['start_time'] == 0) and
```
As discussed offline I don't think we need to change here.

hirenvadalia commented on test/fw/ptl/utils/plugins/ptl_test_runner.py:
```diff
@@ -198,6 +199,9 @@ def _addError(self, test, err):
             self.logger.info('E')
+        if not hasattr(test, 'test'):
+            if not (issubclass(type(test.test), PBSTestSuite)):
+                raise KeyboardInterrupt
```
As discussed offline, PTLTestResult class is only meant for logging results, there should be no errors raise from it.

shilpakodli commented:
@hirenvadalia I have addressed your review comments and also made description more clear and attatched test execution logs.
In test execution, I have just changed framework in order to throw error in CrayUtils() because this issue I found while running tests on Cray Platform and I was able to reproduce same thing on Linux with local changes.

agrawalravi90 commented on test/fw/ptl/utils/plugins/ptl_test_db.py:
```diff
@@ -1725,8 +1726,10 @@ def __create_data(self, test, err=None, status=None):
         testdata['testdoc'] = getattr(_test, '_testMethodDoc', '<unknown>')
-        testdata['start_time'] = getattr(test, 'start_time', 0)
-        testdata['end_time'] = getattr(test, 'end_time', 0)
+        testdata['start_time'] = getattr(test, 'start_time',
```
call datetime.datetime.now() just once, store its value and use it for both start_time and end_time

agrawalravi90 commented on test/fw/ptl/utils/plugins/ptl_test_runner.py:
```diff
@@ -554,11 +555,16 @@ def __get_timeout(self, test):
             method = getattr(test.test, getattr(test.test, '_testMethodName'))
             return getattr(method, TIMEOUT_KEY)
         except:
```
you mentioned in the PR description that this block of code should run when there are any attribute errors, so please catch AttributeError specifically in the except statement.

shilpakodli replied:
@agrawalravi90 In my case, I got AttributeError, but the failure can be anything. That's why I have kept it as general as before.

agrawalravi90 replied:
You could be masking errors if you catch all exceptions and treat them as non-fatal, it's a bad coding practice and discouraged by Python's PEP8 standard: https://www.python.org/dev/peps/pep-0008/ (search for "When catching exceptions")

shilpakodli replied:
@agrawalravi90 "except" in __get_timeout is for the section of the try block, not for any PTL error.  If at all test case fails with any error (IndexError, AttributeError etc), PTL is handiling these and throws proper error as expected. @hirenvadalia Please confirm the same.

hirenvadalia replied:
@shilpakodli I think @agrawalravi90 is right, you only need AttributeError here, have you seen any other errors apart from AttributeError?

shilpakodli replied:
@hirenvadalia and @agrawalravi90 I have made changes. I agree with you.

shilpakodli commented:
> Besides the code comments, your after fix logs show that 2 out of the 4 tests run error out?:
> 
> _error: Failure: IndexError (list index out of range) error: Failure: IndexError (list index out of range) Test cases with failures: Failure.runTest Test suites with failures: Failure run: 4, succeeded: 2, failed: 0, errors: 2, skipped: 0, timedout: 0_

This is because, two of the test suites had CrayUtils() method called inside class "pbs_checkpoint.py" and "pbs_cray_suspend_resume.py". I have made some local changes in order CrayUtils() to throw error and that's why u see two errors in test results which is expected.

shilpakodli commented:
@agrawalravi90 I have addressed your comments and updated test results logs in Description itself.

agrawalravi90 commented:
>  I have made some local changes in order CrayUtils() to throw error and that's why u see two errors in test results which is expected.

Sorry, but could you please explain more, why did you make local changes which are not visible in the PR to make CrayUtils() throw an error? Also, why didn't you run the test suite mentioned in the ticket?

shilpakodli commented:
> > I have made some local changes in order CrayUtils() to throw error and that's why u see two errors in test results which is expected.
> 
> Sorry, but could you please explain more, why did you make local changes which are not visible in the PR to make CrayUtils() throw an error? Also, why didn't you run the test suite mentioned in the ticket?

@agrawalravi90 Thanks for the review. As discussed offline, I have made description more meaningful. Please have a look at it.

shilpakodli commented:
@pbspro-maintainers I have rebased and squashed my commits. Please take this PR forward.
